### PR TITLE
Feature: create a script to ontoportal docker api UI

### PR DESCRIPTION
**Pull request by syphax-bouazzouni**
a copy of https://github.com/ontoportal/ontoportal_docker/pull/6, with the below changes 

## Changes
- [x] Remove old files no more used (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/3fe8fba8f53437f1f8bda02c7a800319b7aaeebf, https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/1a4c0fce6b9184209ee053f679ec3441c289fe2f)
- [x] Make the old `run` script used to run the UI project tests, backward compatible with new ontoportal script (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/e21936a30f28ec7495729300f18b5070e85fe3a0)
- [x] Add unit tests (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/85aea23e0ddc586932f9b841bf3d5ba4d339d7a8)
- [x] Extract three main scripts `bin/run_api.sh`, `bin/run_ui.sh` and `bin/run_cron.sh`, to respectively run/stop the API, UI, and CRON jobs (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/d9e2ac428c3559a5e836e4bcdee830e6dd68c2aa, https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/f7db34994afa1555545a502e029072dd575623b1, https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/d9e2ac428c3559a5e836e4bcdee830e6dd68c2aa)  
- [x] Extract the deploy script to `bin/deploy.sh` file (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/09d0932be88032649e721c5c390fcc26c2c9db91)
- [x] Update the `ontoportal` main script to use the sub-scripts in the bin folder  (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/85e17497ec52be2807fdcee5320e5763ed438435)
- [x] Add Github test CI (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/9af832e8f86b420007661b2fedd0e9d2ccc88acb)
- [ ] Finalize the deployment script 
- [ ] Add the jobs/cron CLI to the `ontoportal` main script, to import/process ontologies, create users,....   
- [x] Add loading animation and ontoportal ASCCI logo  (https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/d6b7595a90ad2b36e2dfc68cd43c9b7c482e1be8, https://github.com/ontoportal-lirmm/ontoportal_docker/pull/4/commits/4cb164f0dbaa602909f7d8c7158e1ef814d1a42b)

### CODE REVIEW

File: ontoportal [hunk 1]
```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env bash
+
+run_api(){
+
+	clone_ongologies_linked_data(){
```
Comment (syphax-bouazzouni) [inline ontoportal]: no need for this now with the new SOLR

File: ontoportal [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	}
+	source .env
+	ontologies_linked_data_link="$ONTOLOGIES_LINKED_DATA_LINK"
+	api_url="$API_URL"
+	if [ -z "$api_url" ] || [ -z "$ontologies_linked_data_link" ]; then
+		echo "[-] Error: Missing required configurations. Please provide both API_URL and ONTOLOGIES_LINKED_DATA_LINK in your .env file"
+		exit 1
+	fi
+	clone_ongologies_linked_data
+	bash_cmd="bundle exec rackup --host 0.0.0.0  --env=development --port 9393"
```
Comment (syphax-bouazzouni) [inline ontoportal]: make this customizable to run the production server or the development server.

File: ontoportal [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  deploy "$2"
+  ;;
+"stop")
+  stop "$2"
+  ;;
+"clean")
+  clean "$2"
+  ;;
+"update")
+  clean "$2"
```
Comment (imadbourouche) [inline ontoportal]: change this to update

File: run [hunk 4]
```diff
@@ -84,42 +91,34 @@ if [ "$PROVISION" = true ]; then
...
     commands=(
-        'docker compose down --volumes > /dev/null 2>&1'
-        'docker compose run --rm  4store bash -c "4s-backend-setup ontoportal_kb" '
-        'docker compose run --rm  ncbo_cron bundle exec rake user:create[admin,admin@nodomain.org,password]'
-        'docker compose run --rm  ncbo_cron bundle exec rake user:adminify[admin]'
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_import --admin-user admin --ontologies $STARTER_ONTOLOGY --from-apikey $API_KEY --from $API_URL'"
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_pull -o $STARTER_ONTOLOGY'"
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_process -o ${STARTER_ONTOLOGY}'"
+        'docker compose  -f docker-compose_api.yml --profile 4store  down --volumes'
+        "bin/run_cron.sh bundle exec rake user:create[admin,admin@nodomain.org,password]'"
```
Comment (imadbourouche) [inline run]: add ' in all the bin/run_cron.sh commands

File: run [hunk 5]
```diff
@@ -84,42 +91,34 @@ if [ "$PROVISION" = true ]; then
...
     echo "Running provisioning ..."
     commands=(
-        'docker compose down --volumes > /dev/null 2>&1'
-        'docker compose run --rm  4store bash -c "4s-backend-setup ontoportal_kb" '
-        'docker compose run --rm  ncbo_cron bundle exec rake user:create[admin,admin@nodomain.org,password]'
-        'docker compose run --rm  ncbo_cron bundle exec rake user:adminify[admin]'
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_import --admin-user admin --ontologies $STARTER_ONTOLOGY --from-apikey $API_KEY --from $API_URL'"
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_pull -o $STARTER_ONTOLOGY'"
-        "docker compose run --rm  ncbo_cron bundle exec 'bin/ncbo_ontology_process -o ${STARTER_ONTOLOGY}'"
+        'docker compose  -f docker-compose_api.yml --profile 4store  down --volumes'
```
Comment (imadbourouche) [inline run]: change this to ./ontoportl clean because there is no docker-compose.yml

File: bin/run_cron.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+		exit 1
+	fi
+
+	source "$env_path"
+
+	docker_run_cmd="docker compose -f docker-compose_api.yml -p ontoportal_docker run  --remove-orphans --rm --name cron-service  --service-ports ncbo_cron bash -c \"$bash_cmd\""
+
+	echo "[+] Starting the CRON"
+	eval "$docker_run_cmd"
+	run_cron "$env_path" "$2"
```
Comment (imadbourouche) [inline bin/run_cron.sh]: remove run_cron

File: ontoportal [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  if [ -z "$1" ]; then
+    SERVICE="ontoportal"
+  else
+    SERVICE="$1"
+  fi
+
+  echo "$SERVICE"
+}
+
+clean() {
```
Comment (imadbourouche) [inline ontoportal:22]: in clean, should  we remove docker compose files for api and ui also ? to each time we start the ontoportal we start it with the new docker compose files

File: bin/run_ui.sh [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  echo "[+] Getting compose file from: $ORGANIZATION_NAME$COMPOSE_UI_FILE_PATH"
+  eval "curl -sS -L https://raw.githubusercontent.com/$ORGANIZATION_NAME$COMPOSE_UI_FILE_PATH -o docker-compose_ui.yml"
+}
+
+status_ok() {
+  curl -sSf http://localhost:3000 >/dev/null 2>&1
+}
+
+log() {
+  docker logs ui-service -f
```
Comment (imadbourouche) [inline bin/run_ui.sh:29]: we can do : `docker exec -it ui-service tail -f log/production.log` like in the run_api

File: ontoportal [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  else
+    echo "Choose which service to display logs API or UI"
+  fi
+}
+
+deploy() {
+  ./bin/deploy.sh "$1"
+}
+
+start() {
```
Comment (imadbourouche) [inline ontoportal]: it's good to call update before starting the script to not fall in the error of old images

File: run [hunk 10]
```diff
@@ -84,42 +91,34 @@ if [ "$PROVISION" = true ]; then
...
-    echo "Waiting for the server to be up..."
-    sleep 1
-    elapsed=$((elapsed + 1))
-done
-if [ $elapsed -ge $timeout ]; then
-    echo "Timed out waiting for the server to be up."
+# Wait for API to be ready (adjust the sleep time accordingly)
+source utils/loading_animation.sh "[+] Waiting for the server http://localhost:9393 to be up..." "http://localhost:9393" 300
+
+if status_ok; then
```
Comment (imadbourouche) [inline run]: status_ok does not exist in the run script

File: run [hunk 11]
```diff
@@ -84,42 +91,34 @@ if [ "$PROVISION" = true ]; then
...
-elapsed=0
-until [ $elapsed -ge $timeout ] || curl -sSf http://localhost:9393 > /dev/null 2>&1; do
-    echo "Waiting for the server to be up..."
-    sleep 1
-    elapsed=$((elapsed + 1))
-done
-if [ $elapsed -ge $timeout ]; then
-    echo "Timed out waiting for the server to be up."
+# Wait for API to be ready (adjust the sleep time accordingly)
+source utils/loading_animation.sh "[+] Waiting for the server http://localhost:9393 to be up..." "http://localhost:9393" 300
```
Comment (imadbourouche) [inline run]: this is being checked in the run_api.sh script

**Review by imadbourouche — CHANGES_REQUESTED** [review]

File: bin/deploy.sh [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	fi
+
+	if [ -z "$1" ]; then
+		SERVICE="ontoportal"
+	else
+		SERVICE="$1"
+	fi
+
+	# Update SERVICE variable in .env file without using sed
+	awk -v SERVICE="$SERVICE" '/^SERVICE=/{sub(/=.*/, "=" SERVICE)} 1' .env >temp && mv temp .env
```
Comment (syphax-bouazzouni) [inline bin/deploy.sh:19]: awk does not work in Mac OS

File: bin/run_ui.sh (hunk 8 above)
Comment (syphax-bouazzouni) [reply bin/run_ui.sh:29]: why ?

File: ontoportal [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+      *)
+        echo "Invalid argument: $arg"
+        return 1
+        ;;
+    esac
+  done
+
+  # Update SERVICE variable in .env
+  SERVICE=$(which_service "$service")
+  sed -i "s/^SERVICE=.*/SERVICE=$SERVICE/" ".env"
```
Comment (syphax-bouazzouni) [inline ontoportal]: sed does not work in MAC OS

**Review by imadbourouche — APPROVED** [review]

**Pull request merged**